packages/apollo-server-core/src/plugin/usageReporting/addTraceRequestStats.ts

```diff
@@ -0,0 +1,428 @@
+        // executed.
+        //
+        // So we don't consider the request invalid in this case. We instead
+        // consider this particular usage as the field being present and provided
+        // null. This is because:
+        // - Schema changes to remove this input field will make this request
+        //   always fail instead of sometimes (hence why we consider it present).
+        // - Schema changes to change the field's type from nullable to
+        //   non-nullable won't care about these stats, since the field's type
+        //   in this case is non-nullable. At first glance, it might seem like
+        //   we'd want to avoid saying the field is provided null when its type
+        //   is non-nullable because it might be confusing. But this just avoids
+        //   the truth about an ugly edge case in spec. More importantly, with
+        //   this information, we can let users know when their requests are
+        //   doing this, as it may be unintentional on their part.
+        inputFieldIsUsedByRequest({
+          trace,
+          inputObjectTypeName: getNamedType(parentInputType).name,
+          inputFieldName: node.name.value,
+          inputFieldTypeName: inputType.toString(),
+          isPresent: true,
+          isNull: node.value.kind === 'NullValue' || (
+            node.value.kind === 'Variable' &&
+            nullVariableNames.has(node.value.name.value)
+          ),
+          isUndefined: node.value.kind === 'Variable' &&
+            undefinedVariableNames.has(node.value.name.value),
+        });
+      },
+      EnumValue(node: EnumValueNode): void {
+        // The operation has been successfully validated by this stage, so the
+        // non-null assertion here is fine. Look at the Kind.ENUM case in
+        // TypeInfo.enter() in graphql-js for why this works.
+        const inputType = typeInfo.getInputType()!;
+        enumValueIsPresentInRequest({
+          trace,
+          enumTypeName: getNamedType(inputType).name,
+          enumValueName: node.value,
+        });
+      },
+    }));
+  } catch (_) {
+    // At the stage of AS when this is run, variables have not been validated
+    // yet, and accordingly the code that traverses variables may throw. In
+    // those cases, we consider the request itself as invalid. Since the point
+    // of collecting these stats is to understand how schema changes affect
+    // valid operation executions, we collect no stats for invalid requests.
+    trace.perInputTypeStat = Object.create(null);
+    trace.perEnumTypeStat = Object.create(null);
+  }
+}
+
+// This code is adapted from coerceInputValue() in graphql-js. Note that we
+// validate the variables here in addition to collecting stats, as we don't
+// want to collect stats for invalid requests.
+function addTraceInputValueStats({
+  trace,
+  inputValue,
+  inputType,
+}: {
+  trace: Trace;
+  inputValue: any;
```

Comment:
I'm very hesitant to allow `undefined` as an `inputValue`. `grapql-js` specifically does not do this in `coerceInputValue()`, presumably because `undefined` isn't so much a value as it is the lack of a key in a map, so it's handled close to where the key exists. One can't encounter an "input value" that's undefined as per spec grammar, but a variable, input field, or argument may be undefined (the specific language used in `coerceVariableValues()` and `coerceArgumentValues()` in spec is whether it "provides a value"). I'm also hesitant because I'm trying to keep the code close enough to `coerceInputValue()` that changes in `coerceInputValue()` could easily be noticed and reflected in this function. I'd rather not deviate from `coerceInputValue()` unless the change doesn't alter the logic significantly, e.g. the `else if` changes you've suggested.